package biomaj;

message DownloadFile {
    required string bank = 1;
    required string session = 2;

    required string local_dir = 3;

    optional int32 timeout_download = 4;

    enum Protocol {
        FTP = 0;
        SFTP = 1;
        HTTP = 2;
        HTTPS = 3;
        DIRECTFTP = 4;
        DIRECTHTTP = 5;
        DIRECTHTTPS = 6;
        LOCAL = 7;
    }

    message Param {
        required string name = 1;
        required string value = 2;
    }

    message HttpParse {
        required string dir_line = 1;
        required string file_line = 2;
        required int32 dir_name = 3;
        required int32 dir_date = 4;
        required int32 file_name = 5;
        required int32 file_date = 6;
        optional string file_date_format = 7;
        required int32 file_size = 8;
    }

    enum HTTP_METHOD {
        GET = 0;
        POST = 1;
    }

    message File {
        // Name of the file
        required string name = 1;
        // Location on remote server, defaults to root directory
        optional string root = 2;
        // Save file under different name
        optional string save_as = 3;
        // Specific file url
        optional string url = 4;
        // Remote file date info to set download file with original date
        optional int32 year = 5;
        optional int32 month = 6;
        optional int32 day = 7;
    }

    message RemoteFile {
        repeated File files = 1;
        required Protocol protocol = 2;
        required string server = 3;
        required string remote_dir = 4;
        optional string save_as = 5;
        repeated Param param = 6;
        optional HttpParse http_parse = 7;
        optional HTTP_METHOD http_method = 8 [ default = GET];
        optional string credentials = 9;
    }

    required RemoteFile remote_file = 5;


    message Proxy {
        required string proxy = 1;
        optional string proxy_auth = 2;
    }

    optional Proxy proxy = 6;

}
